# Machine Learning Lab 4


# Linear Regression from scratch

## 24th February 2021

In [3]:
import numpy as np

The Regression function :

In [8]:
class LinearRegression:

    def __init__(self, lr=0.01, n_iters=1000):
        self.lr = lr
        self.n_iters = n_iters
        self.weights = None
        self.bias = None


    def fit(self, X, y):
        '''
        Fit method takes the features `X` and `y` target
        build a linear model and calculate the weights and bias
        dynamically using `dw` and `db` and update the weights
        and bias using gradient descenting. 
        '''
        n_samples, n_features = X.shape

        # initalize the weights and bias
        self.weights = np.zeros(n_features)
        self.bias = 0


        # Start the learn algorithm gradient descenting
        for _ in range(self.n_iters):
            y_predicted = np.dot(X, self.weights) + self.bias

            # Compute the weights and bias dynamically 
            dw = (1 / n_samples) * np.dot(X.T, (y_predicted - y))
            db = (1 / n_samples) * np.sum(y_predicted - y)

            # Update the weights and bias with actual
            self.weights -= self.lr * dw
            self.bias -= self.lr * db


    def predict(self, X):
        y_predicted = np.dot(X, self.weights) + self.bias
        return np.array(y_predicted)

For error calculation

In [9]:
def mean_squared_error(y_true, y_pred):
    return np.mean((y_true - y_pred)**2)

importing for data and dataframe

In [34]:
from sklearn import datasets
from sklearn.model_selection import train_test_split
import pandas as pd

In [94]:
mse_least = {}
lr = 0.001
n_iters = 500
param = 1

for i in range(1,21):
    X, y = datasets.make_regression(n_samples=100, n_features=i, noise=20, random_state=42)
    # Lets split the dataset into 80 training and 20 testing examples
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=5827)
    
    mse_featureWise = {}
    lr = 0.001
    n_iters = 500
    
    for j in range(1,21):
        clf = LinearRegression(lr, n_iters)
        clf.fit(X_train, y_train)
        prediction = clf.predict(X_test)
        mse = mean_squared_error(y_test, prediction)
        mse_featureWise[mse] = [lr,n_iters]
        
        lr = lr+0.001
        n_iters = n_iters + 500
        
    mse_least[i] = [min(mse_featureWise),mse_featureWise[min(mse_featureWise)][0],mse_featureWise[min(mse_featureWise)][1]]

The mimimum mean square is recorded for a particular parameter number, and its corresponding learning rate and number of iterations i.e for the minimum mse is recorded.

In [95]:
df = pd.DataFrame(mse_least)
df = pd.DataFrame(mse_least, index = ['MSE(Minimum)','Learning Rate','Iteration'])
df = df.T

In [96]:
print(df)

    MSE(Minimum)  Learning Rate  Iteration
1     444.178659          0.020    10000.0
2     267.969718          0.020    10000.0
3     366.274741          0.003     1500.0
4     547.514503          0.003     1500.0
5     267.907681          0.020    10000.0
6     416.661736          0.003     1500.0
7     567.408307          0.003     1500.0
8     366.825518          0.010     5000.0
9     347.928085          0.020    10000.0
10    270.409403          0.020    10000.0
11    446.614999          0.005     2500.0
12    614.992648          0.020    10000.0
13    263.259364          0.020    10000.0
14    504.872197          0.004     2000.0
15    654.090726          0.008     4000.0
16    369.689885          0.020    10000.0
17    645.174056          0.020    10000.0
18    218.967862          0.020    10000.0
19    620.478774          0.020    10000.0
20    887.962472          0.007     3500.0


In [97]:
df.sort_values(by=['MSE(Minimum)'], inplace=True)
print(df)

    MSE(Minimum)  Learning Rate  Iteration
18    218.967862          0.020    10000.0
13    263.259364          0.020    10000.0
5     267.907681          0.020    10000.0
2     267.969718          0.020    10000.0
10    270.409403          0.020    10000.0
9     347.928085          0.020    10000.0
3     366.274741          0.003     1500.0
8     366.825518          0.010     5000.0
16    369.689885          0.020    10000.0
6     416.661736          0.003     1500.0
1     444.178659          0.020    10000.0
11    446.614999          0.005     2500.0
14    504.872197          0.004     2000.0
4     547.514503          0.003     1500.0
7     567.408307          0.003     1500.0
12    614.992648          0.020    10000.0
19    620.478774          0.020    10000.0
17    645.174056          0.020    10000.0
15    654.090726          0.008     4000.0
20    887.962472          0.007     3500.0


We can see the minimum error we got was with 18 parameters, learning rate of 0.020 and 10000 iterations